changelog/dmd.importc-pragma-stc.dd

```diff
@@ -0,0 +1,22 @@
+A pragma for ImportC allows to set `nothrow`, `@nogc` or `pure`
+
+The following new pragma for ImportC allows to set default storage
+classes for function declarations:
+```
+#pragma D [storage classes...]
+```
+The storage classes `nothrow`, `nogc` and `pure` are supported.
+Enabling a default storage class affects all function declarations
+after the pragma until it is disabled with another pragma.
+Declarations in includes are also affected. The following example
+enables `@nogc` and `nothrow` for a library:
+
+```
+#pragma D nogc nothrow
+#include <somelibrary.h>
+```
+
+Default storage classes can be disabled again by adding a minus:
+```
+#pragma D -nogc -nothrow
+```
```

Comment:
Why not `pragma D push(nogc, nothrow)` and `pop`? Which fits better with pragmas of other C compiler vendors.

Reply:
would `pop` just remove `nothrow` in that example?

Reply:
Good idea.

Reply:
> would `pop` just remove `nothrow` in that example?

Using `pop` now reverses the last `push`, which can contain multiple storage classes.

Reply:
Oh ok, so you could do `push(nothrow)` and now the situation is nothrow, but not nogc?

Reply:
`push(...)` adds attributes, but other attributes can still be active. `pop` restores the state before the corresponding `push`. Here is an example:
```
#pragma D push(pure)
    void f1(); // pure
    #pragma D push(nothrow, nogc)
        void f2(); // pure, nothrow, @nogc
    #pragma D pop
    void f3(); // pure
#pragma D pop
void f4(); // no attributes
```
This is equivalent to attribute declarations in D:
```
pure
{
    void f1(); // pure
    nothrow @nogc
    {
        void f2(); // pure, nothrow, @nogc
    }
    void f3(); // pure
}
void f4(); // no attributes
```

Reply:
So I find this a bit less useful. For the case where you want default to be nothrow, nogc, but then you want to declare a function that throws, you need to "pop", and then push the nogc attribute, then pop that, and then push the nothrow, nogc combo to get back to where you were.

Reply:
Yes, this use case would be more complicated. The pragma could be combined with the original idea of using minus, like this: `#pragma D push(-nothrow)`. But I'm not sure, how common this would be.

Reply:
I think it would be the most common case to default to `nothrow, nogc` in C code.

Reply:
In any case, my use case is handled by any of these mechanisms. Having *something* is better than nothing.

Reply:
> I think it would be the most common case to default to `nothrow, nogc` in C code.

The obvious exception being functions that do callbacks via a function-pointer parameter.

Reply:
I have added tests for callbacks. Function pointers get the same attributes as functions, which should cover most use cases.